### ISSUE TITLE
Multiple embedding used to generate an image?

### ISSUE THREAD

**Issue by Deaddawn** [oss-roettger/XL-Textual-Inversion#5]
Hi, there. Is it possible that I train two or more embeddings for say, two characters then use them in a single image?

**Comment by Deaddawn** [comment]
I assume I change code here？

**def set_XLembedding(base,emb,token="my"):
    with torch.no_grad():            
        # Embeddings[tokenNo] to learn
        tokens=base.components["tokenizer"].encode(token)
        assert len(tokens)==3, "token is not a single token in 'tokenizer'"
        tokenNo=tokens[1]
        tokens=base.components["tokenizer_2"].encode(token)
        assert len(tokens)==3, "token is not a single token in 'tokenizer_2'"
        tokenNo2=tokens[1]
        embs=base.components["text_encoder"].text_model.embeddings.token_embedding.weight
        embs2=base.components["text_encoder_2"].text_model.embeddings.token_embedding.weight
        assert embs[tokenNo].shape==emb["emb"].shape, "different 'text_encoder'"
        assert embs2[tokenNo2].shape==emb["emb2"].shape, "different 'text_encoder_2'"
        embs[tokenNo]=emb["emb"].to(embs.dtype).to(embs.device)
        embs2[tokenNo2]=emb["emb2"].to(embs2.dtype).to(embs2.device)**

**Comment by oss-roettger** [comment]
Yes, you can use multiple embeddings to generate images! And you do not even have to change the source code. Just generate two or more embeddings and load those to different tokens that could be used at the same time. e.g.:

    load_XLembedding(base,token="my",embedding_file1)   # e.g. embedding_file1 is the sample doll
    load_XLembedding(base,token="it",embedding_file2)   # e.g. embedding_file2 is your dog
    prompt="my with it "                                # then this prompt would generate the doll with your dog

**Comment by Deaddawn** [comment]
> Yes, you can use multiple embeddings to generate images! And you do not even have to change the source code. Just generate two or more embeddings and load those to different tokens that could be used at the same time. e.g.:
> 
> ```
> load_XLembedding(base,token="my",embedding_file1)   # e.g. embedding_file1 is the sample doll
> load_XLembedding(base,token="it",embedding_file2)   # e.g. embedding_file2 is your dog
> prompt="my with it "                                # then this prompt would generate the doll with your dog 
> ```

got it, thanks

**Comment by Deaddawn** [comment]
> Yes, you can use multiple embeddings to generate images! And you do not even have to change the source code. Just generate two or more embeddings and load those to different tokens that could be used at the same time. e.g.:
> 
> ```
> load_XLembedding(base,token="my",embedding_file1)   # e.g. embedding_file1 is the sample doll
> load_XLembedding(base,token="it",embedding_file2)   # e.g. embedding_file2 is your dog
> prompt="my with it "                                # then this prompt would generate the doll with your dog 
> ```

I have tried this. I found that in certain situation, the certain feature in the characters would mixed, any idea how i can get a better results by improving the prompt or some other ways?

**Comment by oss-roettger** [comment]
The Textual Inversion algorithm learns **everything** that is in your input images (e.g. the background as well). Think of an embedding as a 'Macro'-prompt describing your input images. In my experience the best way to learn separately usable embeddings is to use input images of objects on white background.  Then use the parameter:

    prompt_variations=["on white background"]

For instance, if you want to learn a texture, use input images with the texture in the middle only on a white background, otherwise TI would also learn "completely filled with texture" which would be applied to your generated images when using the texture-embedding anywhere in your prompt. 
For more information see the uncountable articles on Textual Inversion , e.g. start with
https://medium.com/@onkarmishra/how-textual-inversion-works-and-its-applications-5e3fda4aa0bc

**Comment by Deaddawn** [comment]
> The Textual Inversion algorithm learns **everything** that is in your input images (e.g. the background as well). Think of an embedding as a 'Macro'-prompt describing your input images. In my experience the best way to learn separately usable embeddings is to use input images of objects on white background. Then use the parameter:
> 
> ```
> prompt_variations=["on white background"]
> ```
> 
> For instance, if you want to learn a texture, use input images with the texture in the middle only on a white background, otherwise TI would also learn "completely filled with texture" which would be applied to your generated images when using the texture-embedding anywhere in your prompt. For more information see the uncountable articles on Textual Inversion , e.g. start with https://medium.com/@onkarmishra/how-textual-inversion-works-and-its-applications-5e3fda4aa0bc

Oh! Thanks a lot, wonderful suggestion, will take a look.
